### PR TITLE
feat: add support for real-time reporting

### LINKED ISSUE #271: Add support for realtime reporting in Stryker dashboard
This issue depends on stryker-mutator/mutation-testing-elements#2434

---

# Realtime Reporting

Realtime reporting locally is doable, but it would also be nice to have this feature available in the Stryker dashboard.

## Implementation

The implementation is potentially similar to the implementations found in the different Strykers, except that the dashboard should have an API endpoint available for pushing mutants to.

The flow will be like this:
1. Upload the initial state to the dashboard.
2. Accept a list of mutant "events" in the API endpoint (to reduce HTTP calls)
3. Using SSE, relay information to the browser and update the report.
4. When the mutation testing process is complete, send a call to a different endpoint.
5. Write the complete report disk (azure).

## Question(s)

1. When sending mutant data to the dashboard API, should a Stryker tool send the data batched (i.e. wait for 1s and then do the request)?

### PR DESCRIPTION
**Pull request by xandervedder**
~~A couple of things left to do:~~

- [x] Redirect to 'normal' report when real-time report is not available.
- [x] Validate mutants according to JSON schema.
- [x] Separate route for uploading real-time reports.
- [x] Finished event removes append blob.
- [x] ~~Investigate blob removal time (and if I can do this in nodeJS).~~ has to be done in Azure (or in new azurestorage package)
- [x] Pass module name to orchestrator

closes #271

### CODE REVIEW

**Comment by xandervedder** [comment]
Big PR, sorry for that. Should be fully functional now.

File: packages/website-backend/src/index.ts [hunk 1]
```diff
@@ -13,9 +14,11 @@ const port = parseInt(util.optionalEnvVar('PORT', '1337'), 10);
 
 async function ensureDatabaseExists() {
   const repositoryMapper = createProjectMapper();
+  const realTimeMutantsBatchingService = new RealTimeMutantsBlobService();
```
Comment (rouke-broersma) [inline packages/website-backend/src/index.ts:17]:
```suggestion
  const realTimeMutantsBlobService= new RealTimeMutantsBlobService();
```

File: packages/website-backend/src/index.ts [hunk 2]
```diff
@@ -13,9 +14,11 @@ const port = parseInt(util.optionalEnvVar('PORT', '1337'), 10);
...
   const mutationTestingReportService = new MutationTestingReportService();
 
   await repositoryMapper.createStorageIfNotExists();
+  await realTimeMutantsBatchingService.createStorageIfNotExists();
```
Comment (rouke-broersma) [inline packages/website-backend/src/index.ts:21]:
```suggestion
  await realTimeMutantsBlobService.createStorageIfNotExists();
```

File: packages/website-backend/src/api/ReportsController.ts [hunk 3]
```diff
@@ -142,4 +141,18 @@ export default class ReportsController {
...
+  private verifyIfResultIsNotAnIncompleteReport(
+    result: MutationScoreOnlyResult | MutationTestResult
+  ) {
+    if (!isMutationTestResult(result)) {
+      return;
+    }
+
+    if (isPendingReport(result)) {
+      throw new BadRequest(
+        'Not allowed to PUT incomplete reports to the non real-time endpoint.'
```
Comment (rouke-broersma) [inline packages/website-backend/src/api/ReportsController.ts:154]:
```suggestion
        'Submitting pending reports to the completed reports endpoint is not allowed.'
```

File: packages/website-backend/src/api/ReportsController.ts [hunk 4]
```diff
@@ -57,15 +58,17 @@ export default class ReportsController {
       throw new Unauthorized(`Provide an "${API_KEY_HEADER}" header`);
     }
     const slug = req.path;
-    const { project, version } = this.parseSlug(slug);
+    const { project, version } = parseSlug(slug);
     await this.apiKeyValidator.validateApiKey(authorizationHeader, project);
     this.verifyRequiredPutReportProperties(result);
+    this.verifyIfResultIsNotAnIncompleteReport(result);
```
Comment (rouke-broersma) [inline packages/website-backend/src/api/ReportsController.ts:64]:
```suggestion
    this.verifyIsCompletedReport(result);
```

File: packages/website-backend/src/api/RealTimeReportsController.ts [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+      projectName: project,
+      version: version,
+      moduleName,
+      realTime: true,
+    };
+    const server = this.#orchestrator.getSseInstanceForProject(id);
+    server.sendFinished();
+
+    this.#blobService.delete(id);
+    this.#reportService.delete(id);
```
Comment (rouke-broersma) [inline packages/website-backend/src/api/RealTimeReportsController.ts:117]: Does this mean it's up to the client to send the completed report again after? This needs to be very explicitly documented. Did we discuss this?

File: packages/website-backend/src/api/RealTimeReportsController.ts [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    const data = await this.#blobService.getEvents(id);
+    const server = this.#orchestrator.getSseInstanceForProject(id);
+    server.attach(res);
+    data.forEach((mutant) => server.sendMutantTested(mutant));
+  }
+
+  @Delete('/*')
+  public async finish(
```
Comment (rouke-broersma) [inline packages/website-backend/src/api/RealTimeReportsController.ts:95]: Should probably just call this delete instead of finish

File: packages/website-backend/src/api/RealTimeReportsController.ts [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    };
+    const server = this.#orchestrator.getSseInstanceForProject(id);
+    await this.#blobService.appendToBlob(id, mutants);
+    mutants.forEach((mutant) => {
+      server.sendMutantTested(mutant);
+    });
+  }
+
+  @Put('/*')
+  public async update(
```
Comment (rouke-broersma) [inline packages/website-backend/src/api/RealTimeReportsController.ts:161]: When is this used?

File: packages/website-backend/src/services/real-time/SseServer.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      this.#clients.delete(c);
+    });
+  }
+
+  #handleRequest(res: ServerResponse) {
+    res.writeHead(200, {
+      'Content-Type': 'text/event-stream',
+      'Cache-Control': 'no-cache',
+      Connection: 'keep-alive',
+      'Access-Control-Allow-Origin': '*',
```
Comment (rouke-broersma) [inline packages/website-backend/src/services/real-time/SseServer.ts:56]: We should set CORS properly for the dashboard

File: packages/website-backend/src/api/RealTimeReportsController.ts (hunk 5 above)
Comment (xandervedder) [reply packages/website-backend/src/api/RealTimeReportsController.ts:117]: I think we did, as the dashboard should work in a similar way as the reference implementation (Stryker.NET)

Comment (rouke-broersma) [reply packages/website-backend/src/api/RealTimeReportsController.ts:117]: It should be pretty easy to submit the final report from the dashboard backend, right? Could you give me an estimate of the work required?

File: packages/website-backend/src/services/real-time/SseServer.ts (hunk 8 above)
Comment (xandervedder) [reply packages/website-backend/src/services/real-time/SseServer.ts:56]: This should be configurable as this server will be reused eventually, but good call. I guess it would be both the acceptance domain and the 'real' domain?

Comment (rouke-broersma) [reply packages/website-backend/src/services/real-time/SseServer.ts:56]: It should probably be configured during deploy, since prod should only accept requests from prod and vice versa.

Comment (xandervedder) [reply packages/website-backend/src/services/real-time/SseServer.ts:56]: I guess an environment variable would make the most sense here.

File: packages/website-backend/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (xandervedder) [reply packages/website-backend/src/index.ts:17]: I believe Dependency Injection doesn't work in the `index.ts` file, maybe that is why?

File: packages/website-backend/src/services/real-time/SseServer.ts (hunk 8 above)
Comment (xandervedder) [reply packages/website-backend/src/services/real-time/SseServer.ts]: I am not sure where the normal env. variables are set, so it won't work until I figure that out.

File: packages/website-backend/src/services/real-time/MutationEventServerOrchestrator.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    return this.#mutationEventServers.get(id)!;
+  }
+
+  #toId(identifier: ReportIdentifier) {
+    return Object.values(identifier).join(';');
+  }
+
+  #createMutationEventServer(id: string): MutationEventServer {
+    const server = new MutationEventServer(
+      new SseServer(createServer(), this.#configuration.cors)
```
Comment (nicojs) [inline packages/website-backend/src/services/real-time/MutationEventServerOrchestrator.ts:32]: Why do we create a new server instead of reusing the existing one? It seems straightforward to inject the server itself: https://tsed.io/docs/factory.html#inject-http-server-or-https-server

File: packages/website-backend/src/services/real-time/SseServer.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+import { SseClient } from './SseClient.js';
+
+export class SseServer extends EventEmitter {
+  public port?: number;
+
+  #server: Server;
+  #cors: string;
+  #clients = new Set<SseClient>();
+
+  constructor(server: Server, cors = '') {
```
Comment (nicojs) [inline packages/website-backend/src/services/real-time/SseServer.ts:15]: Could we make `SseServer` injectable and use DI to inject the existing server here? See previous remark.

File: packages/website-backend/src/services/Configuration.ts [hunk 11]
```diff
@@ -16,5 +17,11 @@ export default class Configuration {
       util.optionalEnvVar('NODE_ENV', 'production') === 'development'),
       (this.baseUrl = util.requiredEnvVar('STRYKER_DASHBOARD_BASE_URL'));
     this.jwtSecret = util.requiredEnvVar('JWT_SECRET');
+
+    if (this.isDevelopment) {
```
Comment (nicojs) [inline packages/website-backend/src/services/Configuration.ts:21]: I think `isDevelopment` is unnecessary; always use `STRYKER_DASHBOARD_BASE_URL` to set cors. This should also work locally since there this env variable is also set..

**Review by nicojs — CHANGES_REQUESTED** [review]
Thanks for this PR.

The biggest open issue is that you create a new web server per SSE request. That is unnecessary; we should be able to use a `/sse` endpoint in the existing webserver. That would scales much better.

File: packages/website-backend/src/services/real-time/MutationEventServerOrchestrator.ts (hunk 9 above)
Comment (xandervedder) [reply packages/website-backend/src/services/real-time/MutationEventServerOrchestrator.ts:32]: Correct, we do make a new server. However we do not start the server (see the `start()` method), instead we 'attach' to the incoming request (see the `attach(...)`) method.

So everytime a `/sse` request comes in, it attaches itself to that request. I think the reason this was done, is because I reused a lot of functionality from the `SseServer` that is in `mutation-testing-elements`.

Comment (xandervedder) [reply packages/website-backend/src/services/real-time/MutationEventServerOrchestrator.ts:32]: Maybe the `handleRequest` method should become a static one? Might make more sense to do that instead.

File: packages/website-backend/src/api/RealTimeReportsController.ts [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+    const { project, version } = Slug.parse(req.path);
+    await this.#apiKeyValidator.validateApiKey(authorizationHeader, project);
+
+    if (!Array.isArray(mutants)) {
+      throw new BadRequest('Please provide an array of mutant-tested events');
+    }
+
+    if (mutants === null || mutants === undefined) {
+      throw new BadRequest('Please provide mutant-tested events');
+    }
```
[the remark this answers is not in the document]
Comment (xandervedder) [reply packages/website-backend/src/api/RealTimeReportsController.ts:140]: Good question. When I remove both checks all seems to go well. It also appears that the testing server converts objects to arrays and `null` to empty arrays (but I am not sure if the actual server does the same).

File: packages/website-backend/src/api/RealTimeReportsController.ts [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+    const errors = this.#reportValidator.validateMutants(mutants);
+    if (errors !== undefined) {
+      throw new BadRequest(`Invalid mutants: ${errors}`);
+    }
+
+    const id = {
+      projectName: project,
+      version,
+      moduleName: undefined,
```
[the remark this answers is not in the document]
Comment (xandervedder) [reply packages/website-backend/src/api/RealTimeReportsController.ts:150]: I think this was an oversight 😅

File: packages/website-backend/src/services/real-time/MutationEventServerOrchestrator.ts (hunk 9 above)
Comment (xandervedder) [reply packages/website-backend/src/services/real-time/MutationEventServerOrchestrator.ts:32]: I think this should just be a map with the slug as the key and a list of connections as the value.

**Comment by xandervedder** [comment]
Implemented feedback. Servers are not created anymore, instead we just keep a Set of `MutationEventSender`, that will each handle one `Response` (which is what we get when we make a GET call to the realtime endpoint).

---

E2E test seems to be failing, will fix that (if I can figure out how to run them locally)

File: packages/data-access/src/utils.ts [hunk 14]
```diff
@@ -16,3 +17,20 @@ export function isStorageError(
...
+}: ReportIdentifier) {
+  const slug = [
+    projectName,
+    version,
+    moduleName,
+    realTime ? 'real-time' : realTime,
+  ]
+    .filter(Boolean)
+    .join('/');
+  return encodeKey(slug);
```
Comment (nicojs) [inline packages/data-access/src/utils.ts:35]: What will happen when I create a branch named `real-time`? This will be a naming clash, right? Wouldn't it be better to use a separate Container for real-time?

File: packages/e2e/spec/report-api.spec.ts [hunk 15]
```diff
@@ -33,7 +33,7 @@ test.describe('Report api', () => {
     test('should accept v2 reports', async ({ baseURL }) => {
       const response = await client.uploadReport(
         simpleReportv2(
-          'github.com/stryker-mutator-test-organization/hello-org',
+          'github.com/stryker-mutator-test-organization/hello-org?realtime=true',
```
Comment (nicojs) [inline packages/e2e/spec/report-api.spec.ts:36]: I think this causes the e2e test to fail. Please create a separate e2e test for realtime.

File: packages/website-backend/src/api/RealTimeReportsController.ts (hunk 5 above)
Comment (nicojs) [reply packages/website-backend/src/api/RealTimeReportsController.ts:118]: We could maybe add a `HTTP POST /api/real-time/finalize` call, that promotes the real-time results to the finished results.

Maybe add that in a separate PR thought.

File: packages/website-backend/src/services/real-time/MutationEventResponseOrchestrator.ts [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  removeResponseHandlers(identifier: ReportIdentifier) {
+    const id = this.#toId(identifier);
+    this.#responseHandlers.delete(id);
+  }
+
+  #toId(identifier: ReportIdentifier) {
+    return Object.values(identifier).join(';');
+  }
```
Comment (nicojs) [inline packages/website-backend/src/services/real-time/MutationEventResponseOrchestrator.ts:33]: This `toId` function is broken in various ways:

```ts
interface ReportIdentifier {
  projectName: string;
  moduleName: string | undefined;
  version: string;
}

const key1: ReportIdentifier = {
  projectName: 'bar',
  moduleName: 'foo',
  version: 'baz'
};

const key2: ReportIdentifier = {
  moduleName: 'foo',
  projectName: 'bar',
  version: 'baz'
}

const key3 = {
  moduleName: 'foo',
  projectName: 'bar',
  version: 'baz',
  qux: 'quux'
};

function toId(identifier: ReportIdentifier) {
  return Object.values(identifier).join(';');
}

// Should all have the same key
console.log(toId(key1)); // "bar;foo;baz" 
console.log(toId(key2)); // "foo;bar;baz"  
console.log(toId(key3)); // "foo;bar;baz;quux" 
```

File: packages/website-backend/src/services/real-time/MutationEventSender.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+  constructor(res: ServerResponse, cors: string) {
+    this.#response = res;
+    this.#response.writeHead(200, {
+      'Content-Type': 'text/event-stream',
+      'Cache-Control': 'no-cache',
+      Connection: 'keep-alive',
+      'Access-Control-Allow-Origin': cors,
+    });
+    this.#response.on('close', () => this.#destroy());
```
Comment (nicojs) [inline packages/website-backend/src/services/real-time/MutationEventSender.ts:15]: I assume this `close` event is raised when the client disconnects. If so, shouldn't the `response` also be removed from the `MutationEventResponseHandler`?

File: packages/website-backend/test/unit/api/controllers/RealTimeReportsController.spec.ts [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+  let deleteBlobStub: sinon.SinonStubbedMember<
+    MutationTestingReportService['delete']
+  >;
+  let project: Project;
+
+  beforeEach(async () => {
+    await PlatformTest.bootstrap(Server)();
+    request = supertest(PlatformTest.callback());
+
+    const dataAccess = PlatformTest.get<DataAccessMock>(DataAccess);
```
Comment (nicojs) [inline packages/website-backend/test/unit/api/controllers/RealTimeReportsController.spec.ts:59]: Why not use the `dataAccess` directly? Instead of copying over all values to separate `*Stub` variables? Seems like a lot of code for little gain.

File: packages/website-backend/test/unit/api/controllers/RealTimeReportsController.spec.ts [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+        .set('X-Api-Key', apiKey)
+        .send([{ id: '1', status: MutantStatus.Killed }])
+        .expect(200);
+
+      expect(responseHandlerForProjectStub.calledOnce).to.be.true;
+      expect(serverStub.sendMutantTested.firstCall.firstArg).to.deep.include({
+        id: '1',
+        status: MutantStatus.Killed,
+      });
+    });
```
Comment (nicojs) [inline packages/website-backend/test/unit/api/controllers/RealTimeReportsController.spec.ts:213]: Missing a test here: `'should not send a mutant tested to a disconnected client.'`.

**Review by nicojs — CHANGES_REQUESTED** [review]

File: packages/website-backend/src/api/RealTimeReportsController.ts (hunk 5 above)
Comment (xandervedder) [reply packages/website-backend/src/api/RealTimeReportsController.ts:118]: I think we should put that in a seperate PR (might also be a nice issue for hacktoberfest 😄)

File: packages/data-access/src/utils.ts (hunk 14 above)
Comment (xandervedder) [reply packages/data-access/src/utils.ts:35]: The following happens:
```js
    it('what happens when real-time is used as a branch name', () => {
      expect(
        toBlobName({
          projectName: 'abc',
          version: 'real-time',
          moduleName: 'def',
          realTime: true
        })
        ).to.eq('abc;real-time;def;real-time') // passes
    });
```

I don't think this will cause any naming clashes, since real-time adds a fourth part to the blobname.

File: packages/website-backend/src/services/real-time/MutationEventSender.ts (hunk 17 above)
Comment (xandervedder) [reply packages/website-backend/src/services/real-time/MutationEventSender.ts:15]: Maybe, right now I remove the `MutationEventResponseHandler` when real-time testing is done:
https://github.com/stryker-mutator/stryker-dashboard/blob/41a529c3713ca473b0db25ae3fbb03f2ac613272/packages/website-backend/src/api/RealTimeReportsController.ts#L116

But I guess it does make sense if a user disconnects, that nothing is sent to that response any more.

File: packages/website-backend/test/unit/api/controllers/RealTimeReportsController.spec.ts (hunk 19 above)
Comment (xandervedder) [reply packages/website-backend/test/unit/api/controllers/RealTimeReportsController.spec.ts:213]: I don't think that's possible to do since we don't have an actual instance of the `MutationEventResponseOrchestrator`. The name of the test above this one is incorrect, will change that.

File: packages/website-backend/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (xandervedder) [reply packages/website-backend/src/index.ts]: We could hooks for this:
https://tsed.io/docs/hooks.html#introduction

But I feel like this is out of scope for this PR.

**Comment by xandervedder** [comment]
Implemented feedback, also added e2e tests for real-time reporting. I think we are good to go now 😄

File: packages/website-backend/src/services/real-time/MutationEventSender.ts (hunk 17 above)
Comment (xandervedder) [reply packages/website-backend/src/services/real-time/MutationEventSender.ts:18]: It now emits an event :)

File: packages/data-access/src/services/RealTimeMutantsBlobService.ts [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+      toBlobName(id)
+    );
+
+    if (data === '') {
+      return [];
+    }
+
+    return data
+      .split('\n')
+      .slice(0, -1)
```
Comment (xandervedder) [inline packages/data-access/src/services/RealTimeMutantsBlobService.ts:62]: Why is this done?

File: packages/website-backend/test/helpers/TestServer.ts [hunk 21]
```diff
@@ -41,6 +42,7 @@ export const config: Configuration = {
   baseUrl: 'baseUrl',
   jwtSecret: 'jwtSecret',
   isDevelopment: true,
+  cors: '*',
```
Comment (xandervedder) [inline packages/website-backend/test/helpers/TestServer.ts:45]: Check if this is correct or not

File: packages/website-frontend/src/app/report/report-page/report-page.component.ts [hunk 22]
```diff
@@ -60,42 +61,56 @@ export class ReportPageComponent
...
+            if (!object?.report) {
+              this.errorMessage = 'Report does not exist';
+              return;
+            }
+
+            this.id = object.report;
+            if (isMutationTestResult(object.report)) {
+              this.mutationTestResult = object.report;
+              if (isPendingReport(object.report)) {
+                this.sse = `/api/real-time/${object.slug.split("/api/reports/")[1]}`;
```
Comment (xandervedder) [inline packages/website-frontend/src/app/report/report-page/report-page.component.ts:104]: This could be unsafe to do, check this.

**Review by xandervedder — COMMENTED** [review]
Last things.

Tested the working of it by pushing multiple projects at the same time to the dashboard. All seems to be working.

File: packages/data-access/src/services/RealTimeMutantsBlobService.ts (hunk 20 above)
Comment (xandervedder) [reply packages/data-access/src/services/RealTimeMutantsBlobService.ts:62]: Looks like this is used to remove the empty row. I should make that more clear or, even better, use filter instead :)

File: packages/website-backend/test/helpers/TestServer.ts (hunk 21 above)
Comment (xandervedder) [reply packages/website-backend/test/helpers/TestServer.ts:45]: This is in a test file, I am okay with this.

**Comment by xandervedder** [comment]
LGTM
